kuka_rsi_hw_interface/src/robot_control_node.cpp

```diff
@@ -22,66 +22,129 @@
+
+  registerParameter<int>(
+    "n_dof_", 6,
+    kroshu_ros2_core::ParameterSetAccessRights{true, true, false, false},
+    [this](int n_dof)
+    {return this->onNDOFChange(n_dof);});
+
+  kuka_rsi_hw_interface_ = std::make_unique<KukaHardwareInterface>(
+    rsi_ip_address_, rsi_port_,
+    n_dof_);
   return SUCCESS;
 }
-
-rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn RobotControlNode::
-on_activate(
-  const rclcpp_lifecycle::State &)
+CallbackReturn RobotControlNode::on_activate(const rclcpp_lifecycle::State &)
 {
-  kuka_rsi_hw_interface_->start();
+  kuka_rsi_hw_interface_->start(
+    joint_state_msg_.position, joint_command_msg_->position,
+    initial_joint_pos_);
 
-  control_thread_ = std::thread(&RobotControlNode::Controlloop, this);
+  joint_state_publisher_->on_activate();
+  joint_state_publisher_->publish(joint_state_msg_);
+
+  control_thread_ = std::thread(&RobotControlNode::ControlLoop, this);
   return SUCCESS;
 }
 
-rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn RobotControlNode::
-on_deactivate(
-  const rclcpp_lifecycle::State &)
+CallbackReturn RobotControlNode::on_deactivate(const rclcpp_lifecycle::State &)
 {
-  kuka_rsi_hw_interface_->stop();
-  if (control_thread_.joinable()) {control_thread_.join();}
+  kuka_rsi_hw_interface_->stop(joint_pos_correction_deg_);
+  joint_state_publisher_->on_deactivate();
+  if (control_thread_.joinable()) {
+    control_thread_.join();
+  }
 
   return SUCCESS;
 }
 
-rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn RobotControlNode::
-on_cleanup(
-  const rclcpp_lifecycle::State &)
+CallbackReturn RobotControlNode::on_cleanup(const rclcpp_lifecycle::State &)
 {
-  kuka_rsi_hw_interface_->cleanup();
+  kuka_rsi_hw_interface_ = nullptr;
   return SUCCESS;
 }
 
-void RobotControlNode::Controlloop()
+void RobotControlNode::ControlLoop()
 {
   while (kuka_rsi_hw_interface_->isActive()) {
-    if (!kuka_rsi_hw_interface_->read()) {
+    std::unique_lock<std::mutex> lock(m_);
+    if (!kuka_rsi_hw_interface_->read(joint_state_msg_.position)) {
       RCLCPP_ERROR(get_logger(), "Failed to read state from robot. Shutting down!");
       rclcpp::shutdown();
       return;
     }
-    kuka_rsi_hw_interface_->write();
+    joint_state_publisher_->publish(joint_state_msg_);
+
+    cv_.wait(lock);   // wait for new command -> (Marton) is it really needed?
+
+    for (size_t i = 0; i < n_dof_; i++) {
+      joint_pos_correction_deg_[i] = (joint_command_msg_->position[i] - initial_joint_pos_[i]) *
+        RobotControlNode::R2D;
+    }
+
+    kuka_rsi_hw_interface_->write(joint_pos_correction_deg_);
   }
 }
-}  // namespace kuka_rsi_hw_interface
 
+void RobotControlNode::commandReceivedCallback(sensor_msgs::msg::JointState::SharedPtr msg)
+{
+  std::lock_guard<std::mutex> lock(m_);
+  joint_command_msg_ = msg;
+  cv_.notify_one();
+}
+
+bool RobotControlNode::onRSIIPAddressChange(const std::string & rsi_ip_address)
```

Comment:
maybe check IP validity?

Reply:
I would declare all parameters in constructor, as for now, they can't be changed in unconfigured state